File: backend/eyes.py
"""
严格真眼判定

真眼条件：
  1. 空点
  2. 4 正交邻全属同一连通群
  3. 对角：边/角全同色；内部 ≥3 同色
"""

from typing import List, Optional, Set, Tuple

from board import Board, EMPTY

DIAG_DIRS = [(-1, -1), (1, -1), (-1, 1), (1, 1)]


def get_target_group(board: Board, coord: Tuple[int, int]) -> Optional[dict]:
    """取 coord 所在群信息。返回 {color, group, group_set, libs, lib_count} 或 None。"""
    tx, ty = coord
    color = board.get(tx, ty)
    if color == EMPTY:
        return None
    group, libs = board.group_and_libs(tx, ty)
    group_set = {gy * board.size + gx for gx, gy in group}
    return {
        "color": color,
        "group": group,
        "group_set": group_set,
        "libs": libs,
        "lib_count": len(libs),
    }


def is_eye_of_group(board: Board, x: int, y: int,
                    color: int, group_set: Set[int]) -> bool:
    size = board.size
    i = y * size + x
    if board.grid[i] != EMPTY:
        return False
    ortho_count = 0
    if y > 0:
        ortho_count += 1
        if (i - size) not in group_set: return False
    if y < size - 1:
        ortho_count += 1
        if (i + size) not in group_set: return False
    if x > 0:
        ortho_count += 1
        if (i - 1) not in group_set: return False
    if x < size - 1:
        ortho_count += 1
        if (i + 1) not in group_set: return False
    on_edge = ortho_count < 4
    diag_in = 0
    diag_same = 0
    for dx, dy in DIAG_DIRS:
        nx, ny = x + dx, y + dy
        if not board.in_bounds(nx, ny): continue
        diag_in += 1
        if board.get(nx, ny) == color: diag_same += 1
    if on_edge:
        return diag_same == diag_in
    return diag_same >= 3


def count_real_eyes(board: Board, target: Optional[dict]) -> int:
    if target is None:
        return 0
    color = target["color"]
    group_set = target["group_set"]
    libs = target["libs"]
    size = board.size
    eyes = 0
    for k in libs:
        x = k % size
        y = k // size
        if is_eye_of_group(board, x, y, color, group_set):
            eyes += 1
            if eyes >= 2:
                return 2
    return eyes

File: eyes.js
// ============================================================
// 严格真眼判定（绑定特定目标群）
// ============================================================
//
// 一个空点 P 是"color 方的真眼"当且仅当：
//   1. P 为空点
//   2. P 的 4 个正交邻（存在的）全部属于指定 group（同一连通块）
//   3. 对角线条件：
//      - P 在边/角：所有对角邻（存在的）必须为 color
//      - P 在内部：4 对角中至少 3 个为 color
//
// 返回一个群内的真眼总数（饱和到 2，因为 2 就是活）。
// ============================================================

const DIAG_DIRS = [[-1, -1], [1, -1], [-1, 1], [1, 1]];

// 取得 targetCoord 所在群（及其坐标集合），不存在则返回 null
function getTargetGroup(board, targetCoord) {
  const [tx, ty] = targetCoord;
  const color = board.get(tx, ty);
  if (color === E) return null;
  const info = board.groupAndLibs(tx, ty);
  return {
    color,
    group: info.group,
    groupMask: info.groupMask,
    libs: info.libs,
    libArray: info.libArray,
  };
}

// 判断 (x,y) 空点是否为 target 群的真眼
// groupMask: Uint8Array，1 表示该索引属于目标群
function isEyeOfGroup(board, x, y, color, groupMask) {
  const size = board.size;
  const i = y * size + x;
  if (board.grid[i] !== E) return false;

  // 正交邻：全部必须属于本群
  // 展开判断以避免分配临时数组
  let orthoCount = 0;
  if (y > 0) {
    const ni = i - size;
    if (!groupMask[ni]) return false;
    orthoCount++;
  }
  if (y < size - 1) {
    const ni = i + size;
    if (!groupMask[ni]) return false;
    orthoCount++;
  }
  if (x > 0) {
    const ni = i - 1;
    if (!groupMask[ni]) return false;
    orthoCount++;
  }
  if (x < size - 1) {
    const ni = i + 1;
    if (!groupMask[ni]) return false;
    orthoCount++;
  }
  const onEdge = orthoCount < 4;

  // 对角检测
  let diagInBounds = 0;
  let diagSame = 0;
  for (let d = 0; d < 4; d++) {
    const nx = x + DIAG_DIRS[d][0];
    const ny = y + DIAG_DIRS[d][1];
    if (nx < 0 || nx >= size || ny < 0 || ny >= size) continue;
    diagInBounds++;
    if (board.grid[ny * size + nx] === color) diagSame++;
  }

  if (onEdge) return diagSame === diagInBounds;
  return diagSame >= 3;
}

// 统计指定群的真眼数（饱和到 2）
function countGroupRealEyes(board, target) {
  if (!target || !target.groupMask) return 0;
  const { color, groupMask, libArray } = target;
  const size = board.size;
  let eyes = 0;
  for (let i = 0; i < libArray.length; i++) {
    const k = libArray[i];
    const x = k % size;
    const y = (k - x) / size;
    if (isEyeOfGroup(board, x, y, color, groupMask)) {
      eyes++;
      if (eyes >= 2) return 2;
    }
  }
  return eyes;
}
